### PR TITLE
CFE-3575/CFE-3576 Added tests for custom promise libraries

### PR DESCRIPTION
**Pull request by Lex-2008**
----

#

### CODE REVIEW

**Comment by Lex-2008** [comment]
Travis is green. @cf-bottom build Jenkins please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6180/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6180/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6180/

**Comment by Lex-2008** [comment]
with exotics: https://ci.cfengine.com/job/pr-pipeline/6181/

File: Makefile.am [hunk 1]
```diff
@@ -56,7 +56,7 @@ TAR_OPTIONS = --owner=0 --group=0
 export TAR_OPTIONS
 
 
-EXTRA_DIST = ChangeLog INSTALL README.md LICENSE CFVERSION
+EXTRA_DIST = ChangeLog INSTALL README.md LICENSE CFVERSION docs/custom_promise_types/cfengine.sh
```
Comment (vpodzime) [inline Makefile.am]: Once again I have to say that these things shouldn't be under *docs/*. What about *misc/custom_promise_types*?

File: tests/acceptance/30_custom_promise_types/05_cp_using_library.cf [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+{
+      dirlinks => "delete";
+      rmdirs   => "true";
+}
+
+#######################################################
+
+promise agent example
+{
+    interpreter => "/bin/bash";
```
Comment (vpodzime) [inline tests/acceptance/30_custom_promise_types/05_cp_using_library.cf:44]: Is this going to work on all exotics?

File: tests/acceptance/30_custom_promise_types/05_cp_using_library.cf [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+bundle agent test
+{
+  meta:
+    "description" -> { "CFE-3575" }
+      string => "Test a simple promise module which uses library";
+
+  example:
+    cfengine::
```
Comment (vpodzime) [inline tests/acceptance/30_custom_promise_types/05_cp_using_library.cf:55]: Why the class guard?

File: tests/acceptance/30_custom_promise_types/06_cp_with_classes.cf [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+body delete init_delete
+{
+      dirlinks => "delete";
+      rmdirs   => "true";
+}
+
+#######################################################
+
+promise agent example
```
Comment (vpodzime) [inline tests/acceptance/30_custom_promise_types/06_cp_with_classes.cf:41]: `promise agent cp` (or change the examples in modules to use the `example` promise type)

File: tests/acceptance/30_custom_promise_types/07_python_library.cf [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+bundle common python_check
+{
+  classes:
+    "python_f_string_support" expression => returnszero("/usr/bin/python3 -c 'f\"string\"'", "useshell");
+}
+
+bundle agent init
+{
+  meta:
+      "test_skip_unsupported" string => "!python_f_string_support";
```
Comment (vpodzime) [inline tests/acceptance/30_custom_promise_types/07_python_library.cf:24]: So where does this test run? Is it even worth having?

File: Makefile.am (hunk 1 above)
Comment (vpodzime) [reply Makefile.am]: And then this should be added to `EXTRA_DIST` in the particular folder, not in the top-level *Makefile.am*.

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise.

File: tests/acceptance/30_custom_promise_types/05_cp_using_library.cf [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+body delete init_delete
+{
+      dirlinks => "delete";
+      rmdirs   => "true";
+}
+
+#######################################################
+
+promise agent example
```
Comment (craigcomstock) [inline tests/acceptance/30_custom_promise_types/05_cp_using_library.cf:42]: change to `promise agent cp` as vrata mentioned further down?

File: tests/acceptance/30_custom_promise_types/07_python_library.cf (hunk 5 above)
Comment (Lex-2008) [reply tests/acceptance/30_custom_promise_types/07_python_library.cf:24]: Yes, it runs on all 7 platforms marked as red in this testrun: https://ci.cfengine.com/job/testing-pr/13238/ (except for i386_linux_redhat_5 which failed for a different reason).

File: Makefile.am (hunk 1 above)
Comment (Lex-2008) [reply Makefile.am]: I don't have a strong opinion on this, but what would the person who put them there, in the first place, say? @olehermanse ?

**Review by vpodzime — COMMENTED** [review]
Looks good to me apart from the location of the libraries and the path to *bash*.

File: tests/acceptance/30_custom_promise_types/05_cp_using_library.cf (hunk 2 above)
Comment (Lex-2008) [reply tests/acceptance/30_custom_promise_types/05_cp_using_library.cf:44]: well, the test passes :) So I believe it means the file `/bin/bash` exists on all exotics, and is capable of parsing shell scripts.

**Comment by Lex-2008** [comment]
@cf-bottom build exotics please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6244/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6244/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6244/

File: Makefile.am (hunk 1 above)
Comment (olehermanse) [reply Makefile.am]: They are in `docs/custom_promise_types/` just to have everything in 1 place. I think the libraries should be moved to `masterfiles/modules/`

**Comment by Lex-2008** [comment]
https://ci.cfengine.com/job/pr-pipeline/6264/ :green_heart:

File: misc/Makefile.am [hunk 7]
```diff
@@ -9,7 +9,9 @@ EXTRA_DIST= init.d/cfengine3.in \
 	systemd/cf-postgres.service.in \
 	systemd/cf-runalerts.service.in \
 	systemd/cf-serverd.service.in \
-	systemd/cfengine3.service.in
+	systemd/cfengine3.service.in \
+  custom_promise_types/cfengine.sh \
```
Comment (vpodzime) [inline misc/Makefile.am:13]: Wrong indentation.

**Review by vpodzime — COMMENTED** [review]
I think the documentation for custom promise types should stay in *docs*. Only the libraries should be moved under *misc*.

**Comment by Lex-2008** [comment]
ok! What about promise modules? Some of them use these libs, others don't.

**Comment by olehermanse** [comment]
@Lex-2008 They are examples and should stay in docs

**Comment by Lex-2008** [comment]
right, and in order to run them, you'll have to copy the library from misc to docs

**Comment by olehermanse** [comment]
> right, and in order to run them, you'll have to copy the library from misc to docs

@Lex-2008 no, you put them all in `/var/cfengine/masterfiles/modules/promises/`, which makes them land in the correct `/var/cfengine/modules/promises/`.

**Comment by Lex-2008** [comment]
indeed!

**Review by craigcomstock — COMMENTED** [review]
looks good to me other than vrata's comments. :+1:

**Review by vpodzime — APPROVED** [review]
ACK

**Comment by vpodzime** [comment]
@cf-bottom jenkins, please

**Review by craigcomstock — APPROVED** [review]

**Comment by vpodzime** [comment]
@olehermanse feel free to merge this if it looks good to you.

**Pull request merged**